### PR TITLE
ncurses: Build standard terminfo dirs into static ncurses

### PR DESCRIPTION
**Pull request by NorfairKing**
## Description of changes

This builds terminfo directories into ncurses when building statically, even on non-static variants of nixpkgs.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/ncurses/default.nix
Comment (midchildan) [inline pkgs/development/libraries/ncurses/default.nix:42]: I think it can be simplified to this:

```suggestion
  ] ++ lib.optionals (stdenv.hostPlatform.isUnix && enableStatic) [
```

I don't think the package would build if `stdenv.hostPlatform.isStatic == true && enableStatic == false`.

Comment (NorfairKing) [reply pkgs/development/libraries/ncurses/default.nix:42]: @midchildan Does `enableStatic` get activated automatically on `stdenv.hostPlatform.isStatic`?
Or do I need to change the default value for `enableStatic` to `stdenv.hostPlatform.isStatic`?

Comment (midchildan) [reply pkgs/development/libraries/ncurses/default.nix:42]: The default value for `enableStatic` already is `stdenv.hostPlatform.isStatic`, so no further changes should be necessary.

https://github.com/NixOS/nixpkgs/blob/742eb483001595f27be20c2eee501be8623150fe/pkgs/development/libraries/ncurses/default.nix#L8

**Comment by NorfairKing** [comment]
Done

**Comment by NorfairKing** [comment]
@midchildan Anything else I need to do before we can merge this? I'd like to get it into 24.05.

**Comment by midchildan** [comment]
`passthru.tests` is failing, but this would probably be fixed with #311069

**Review by midchildan — APPROVED** [review]

**Comment by midchildan** [comment]
@NorfairKing You'll need to get a committer to merge it. Since I reviewed this PR, you can post a link to the "PRs already reviewed" Discourse thread to give it some attention.

https://discourse.nixos.org/t/prs-already-reviewed/2617

**Comment by JohnRTitor** [comment]
> I'd like to get it into `24.05`.

Probably won't happen unless backported. The last staging-next iteration was merged a few days ago. 

I wonder if this should be targeting master branch though, it doesn't cause _that many_ rebuilds.

**Comment by NorfairKing** [comment]
I could backport it once this is merged.

**Comment by doronbehar** [comment]
I don't understand, how did you find out you need to apply this change?

**Comment by NorfairKing** [comment]
@doronbehar I was statically linking against musl with `pkgsMusl`.

**Comment by doronbehar** [comment]
> @doronbehar I was statically linking against musl with `pkgsMusl`.

Exactly what expression did you try to evaluate? Could you share it please?

**Comment by NorfairKing** [comment]
@doronbehar you can check here: https://github.com/NorfairKing/smos/blob/8c7f97b3aba304e6d0ac6d36493f9c22600a4572/nix/overlay.nix#L54
and also the terminfo section here: https://cs-syd.eu/posts/2024-04-20-static-linking-haskell-nix

**Comment by NorfairKing** [comment]
This PR has already been approved. I just need someone to press merge, and I don't know what anyone is waiting for .. ?
This process could use some transparency :(

**Comment by doronbehar** [comment]
> @doronbehar you can check here: [NorfairKing/smos@`8c7f97b`/nix/overlay.nix#L54](https://github.com/NorfairKing/smos/blob/8c7f97b3aba304e6d0ac6d36493f9c22600a4572/nix/overlay.nix#L54) and also the terminfo section here: [cs-syd.eu/posts/2024-04-20-static-linking-haskell-nix](https://cs-syd.eu/posts/2024-04-20-static-linking-haskell-nix)

It's a bit hard for me to dive into the context there, but I'm taking from here that basically the issue with the `terminfo` dirs is reproducible with (if you are inside a Nixpkgs branch):

```
nix build -L --impure --expr 'with (builtins.getFlake "'$PWD'").legacyPackages.${builtins.currentSystem}; ncurses.override {enableStatic = true;}'
```

I asked that because I wanted to claim that if you'd run instead:

```
nix build -L --impure --expr 'with (builtins.getFlake "'$PWD'").legacyPackages.${builtins.currentSystem}; pkgsStatic.pkgsMusl.ncurses'
```

You'd get the package you are interested in, without the `terminfo` dirs issue.

This case makes me think that in general we shouldn't add `enableStatic` arguments to packages, but instruct people to use `pkgsStatic.${pkg}`, because they really mean the same.

> This PR has already been approved. I just need someone to press merge, and I don't know what anyone is waiting for .. ? This process could use some transparency :(

I might merge it, I just want to make sure that my observations are correct. I'm sorry for you were waiting so long. Your PR is not trivial although it seems simple, which makes people afraid to break packages due to it. I'm opening a discourse thread about this issue in general.

I'm sorry I wasn't transparent in my intentions [when I commented here for the first time](https://github.com/NixOS/nixpkgs/pull/311411#issuecomment-2132185481), I hope now that my concerns are clear.

**Comment by doronbehar** [comment]
BTW this doesn't have to target `staging` branch, as the number of rebuilds is small because it touches only the static `ncurses`.

**Comment by NorfairKing** [comment]
CC @nh2 who's done all the work you're suggesting removing.
IIRC he knows about why we use `pkgsMusl` instead of `pkgsStatic` for static linking.

**Comment by doronbehar** [comment]
> CC @nh2 who's done all the work you're suggesting removing.

The work that *I* suggest removing? Are you talking about removing the `enableStatic` arguments?

**Comment by NorfairKing** [comment]
> The work that I suggest removing? Are you talking about removing the enableStatic arguments?

Yes, the discourse page https://discourse.nixos.org/t/remove-all-enablestatic-arguments-in-nixpkgs-and-instruct-people-to-use-pkgsstatic/46092 seems to suggest removing `enableStatic`, right?
Did I read that wrong?

This would make impossible to link statically in a nixpkgs "version" like `pkgsMusl`, which is what @nh2 has made possible in many cases, see https://github.com/NixOS/nixpkgs/issues/43795

This PR is just following up on the years of effort that went into that, so I'm pinging him to weigh in.

**Comment by doronbehar** [comment]
> seems to suggest removing `enableStatic`, right?
> Did I read that wrong?

Partly wrong yes. I'm suggesting to remove `enableStatic`, but rely instead upon `stdenv.hostPlatform.isStatic` - so that `pkgsStatic.ncurses` and/or `pkgsStatic.pkgsMusl.ncurses` will make the expression add the `configureFlags` you are missing.

> This would make impossible to link statically in a nixpkgs "version" like `pkgsMusl`, which is what @nh2 has made possible in many cases, see #43795

OK I wasn't aware of this effort. I had in the past good experience with `pkgsStatic.pkgsMusl` and `pkgsMusl.pkgsStatic`, so that's why I suggested you to try out `pkgsStatic.pkgsMusl.ncurses`, and perhaps also try (it could be equivalent) `pkgsMusl.pkgsStatic.ncurses`, as I'm curious to know how will it work for you - that's what I asked you to try to use in [my long comment](https://github.com/NixOS/nixpkgs/pull/311411#issuecomment-2134340753). Please try out these attribute as an alternative :pray:.

**Comment by NorfairKing** [comment]
@doronbehar I'm all for improvements but can we please make them after merging this fix ?

**Comment by doronbehar** [comment]
OK I understand now the subtle difference between `enableStatic` and using `pkgsStatic` :) Sorry for holding this up, and thank you for insisting you are correct :rocket: .

**Comment by doronbehar** [comment]
Aish you could have target branch `master`, and could have get this update faster to your channels / flakes.

**Comment by NorfairKing** [comment]
@doronbehar thank you for the kind communication :yellow_heart:

**Comment by NorfairKing** [comment]
@doronbehar Some more background info:
https://github.com/NixOS/nixpkgs/issues/61575

**Pull request merged**
